src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -482,14 +439,53 @@ private void ProcessRenderQueue()
             }
         }
 
-        private void InvokeRenderCompletedCalls(ArrayRange<RenderTreeDiff> updatedComponents)
+        private async void InvokeRenderCompletedCalls(ArrayRange<RenderTreeDiff> updatedComponents)
         {
+            List<Task> batch = null;
             var array = updatedComponents.Array;
             for (var i = 0; i < updatedComponents.Count; i++)
             {
-                // The component might be rendered and disposed in the same batch (if its parent
-                // was rendered later in the batch, and removed the child from the tree).
-                GetOptionalComponentState(array[i].ComponentId)?.NotifyRenderCompleted();
+                var componentState = GetOptionalComponentState(array[i].ComponentId);
+                if (componentState != null)
+                {
+                    // The component might be rendered and disposed in the same batch (if its parent
+                    // was rendered later in the batch, and removed the child from the tree).
+                    var task = componentState.NotifyRenderCompletedAsync();
+
+                    // We want to avoid allocations per rendering. Avoid allocating a state machine or an accumulator
+                    // unless we absolutely have to.
+                    if (task.IsCompleted)
+                    {
+                        if (task.Status == TaskStatus.RanToCompletion || task.Status == TaskStatus.Canceled)
+                        {
+                            // Nothing to do here.
+                            continue;
+                        }
+                        else if (task.Status == TaskStatus.Faulted)
+                        {
+                            HandleException(task.Exception);
+                            continue;
+                        }
+                    }
+
+                    // The Task is incomplete.
+                    // Queue up the task and we can inspect it later.
+                    batch = batch ?? new List<Task>();
+                    batch.Add(task);
+                }
+            }
+
+            if (batch != null)
+            {
+                var aggregateTask = Task.WhenAll(batch);
+                try
+                {
+                    await aggregateTask;
+                }
+                catch when (!aggregateTask.IsCanceled)
```

Comment:
I think you should do this per individual task, not on the Task.WhenAll as you want to handle exceptions from individual components